lab2/two_drones_pkg/src/plots_publisher_node.cpp

```diff
@@ -1,152 +1,172 @@
-          *                 needs to be plotted (given)
-          *   - parent->tf_listener, a tf::TransformListener object (given)
-          *   - transform, the tf::transform object that needs to be populated
-          *
-          * HINT: use "lookupTransform", see
-          *  http://wiki.ros.org/tf/Tutorials/Writing%20a%20tf%20listener%20%28C%2B%2B%29#The_Code
-	        *
-	        * NOTE: only using "lookupTransform", you might see a lot of errors in your terminal!
-          *       Unlike other topic types - which are typically received on a callback basis -
-          *       \tf can be polled at any time and therefore "lookupTransform" might fail 
-          *       due to timing issues. To fix it, consider the method "waitForTransform" with 
-          *       a small wait time.    
-          */
-
-         // ~~~~~~~~~~~~~~~~~~~~~~~~  END OF EDIT SECTION  ~~~~~~~~~~~~~~~~~~~~~~~~~
-         while(poses.size() >= buffer_size)
-            poses.pop_front();
-
-         tf::Vector3 pos(transform.getOrigin());
-         geometry_msgs::Point tmp;
-         tmp.x = pos.x(); tmp.y = pos.y(); tmp.z = pos.z();
-         poses.push_back(tmp);
-     } catch (tf::TransformException ex) {
-         ROS_ERROR("%s",ex.what());
-     }
-   }
+class PlotsPublisherNode {
+  ros::Time startup_time;
+  ros::Timer heartbeat;
+  ros::NodeHandle nh;
+  ros::Publisher markers_pub;
+  tf2_ros::TransformListener tf_listener;
+  tf2_ros::Buffer tf_buffer;
+  int num_trails;
+
+  class TrajTrail {
+    PlotsPublisherNode* parent;
+    static int id;
+    std::list<geometry_msgs::Point> poses;
+    std::string ref_frame, dest_frame;
+    size_t buffer_size;
+
+    std::string ns;
+    float r, g, b, a;
+
+    visualization_msgs::Marker marker_out;
+
+    void update() {
+      // NOTE: you need to populate this transform
+      geometry_msgs::TransformStamped transform;
+      try {
+        // ~~~~~~~~~~~~~~~~~~~~~~  BEGIN OF EDIT SECTION  ~~~~~~~~~~~~~~~~~~~~~~~~~
+
+        /* The transform object needs to be populated with the most recent
+         * transform from ref_frame to dest_frame as provided by tf.
+
+         * Relevant variables in this scope:
+         *   - ref_frame, the frame of reference relative to which the trajectory
+         *                needs to be plotted (given)
+         *   - dest_frame, the frame of reference of the object whose trajectory
+         *                 needs to be plotted (given)
+         *   - parent->tf_buffer, a tf2_ros::Buffer object (given)
+         *   - transform, the geometry_msgs::TransformStamped object that needs to be populated
+         *
+         * HINT: use "lookupTransform", see
+         * https://wiki.ros.org/tf2/Tutorials/Writing%20a%20tf2%20listener%20%28C%2B%2B%29#TheCode
+         */
+
+        // *** FILL IN ***
+
+        // ~~~~~~~~~~~~~~~~~~~~~~~~  END OF EDIT SECTION  ~~~~~~~~~~~~~~~~~~~~~~~~~
+        while (poses.size() >= buffer_size) {
```

Comment:
Will buffer_size be zero? If so pop_front() is undefined. Maybe we want to add a check for buffer_size.